### DISCUSSION TITLE
SQLAlchemy test hangs with Ingres

### DISCUSSION THREAD

**Discussion by hab6** [sqlalchemy/sqlalchemy#11295] [Usage Questions]
I am encountering a deadlock running a particular test of the SQLAlchemy test suite with an Ingres 11.2 backend and the [Ingres dialect](https://github.com/ActianCorp/sqlalchemy-ingres). The Ingres dialect uses pyodbc for database connections.

The deadlock occurs with this command to run the specific test `test_save_update`:

    pytest --db ingres_odbc test\orm\inheritance\test_basic.py -k "test_save_update"

Reviewing the relevant operations from an ODBC trace shows:

    Connection 1
    SELECT base.id AS base_id, base.version_id AS base_version_id, base.value AS base_value,
    base.discriminator AS base_discriminator FROM base WHERE base.id = ?"

    (No COMMIT or ROLLBACK occurs)

    Connection 2
    UPDATE base SET version_id=? WHERE base.id = ? AND base.version_id = ?

From the ODBC trace and other testing with Ingres apart from SQLAlchemy, the hang is expected and occurs because:
 - The Ingres DBMS isolation level is set to SERIALIZABLE.
 - There are two connections and each have set AUTOCOMMIT to OFF.
 - A SELECT is performed on one connection without being followed by a COMMIT or ROLLBACK.
 - An UPDATE is attempted on the other connection, which hangs waiting for the lock to clear from the first connection.

A couple more things worth noting:
- The ODBC trace shows a COMMIT and ROLLBACK occur after the earlier INSERT statements that populated the tables, but no COMMIT or ROLLBACK occurs in between the SELECT on connection 1 and the UPDATE on connection 2.
- I have tried other isolation levels for the Ingres database without success: UNCOMMITTED READ (transaction error updating a READ ONLY database), COMMITTED READ (~deadlock~ error on id mismatch followed by deadlock on second session flush), REPEATABLE READ (deadlock)
- If the same test is run against a PostreSQL or SQLite backend, no hang occurs.

**My Questions:** 
Why are two connections used by SQLAlchemy/pyodbc and no COMMIT or ROLLBACK executed in between the SELECT and UPDATE statements? Am I missing something simple? Is this a flaw in the Ingres dialect?
 
#### Environment
Actian Ingres 11.2 on Windows 10
Python 3.10.7
mock              5.1.0
packaging         24.0
pip               24.0
pluggy            1.4.0
psycopg2          2.9.9
pyodbc            5.1.0
pypyodbc          1.3.6
pytest            8.1.1
setuptools        69.2.0
SQLAlchemy        2.0.29.dev0
sqlalchemy-ingres 0.0.7.dev1
tox               4.14.2

**Comment by CaselIT** [comment]
Hi,

Note that it's not expected for 3rd party to pass the complete tests of sqlalchemy, just the test suite that's in `sqlalchemy.testing`.

Does changing the isolation level to read-commited fixes this?
Note that tests seems to explicitly test two connections at the same time, but it may be assuming a particular serialization level

**Comment by hab6** [comment]
Thanks for the response.

With read-committed, the test raises the `orm_exc.StaleDataError` assertion and reports:

    b"Instance '<Sub at 0x2369b63c220>' has version id '1' which does not match database-loaded version id '2'."

Continuing on, it hangs on the last statement in the script, when calling `sess2.flush`.

Regarding the test suite, I usually run the tests under the [test](https://github.com/sqlalchemy/sqlalchemy/tree/main/test) directory and exclude some things, like the dialects folder.

In mentioning `sqlalchemy.testing` are you referring to the tests under this [testing](https://github.com/sqlalchemy/sqlalchemy/tree/main/lib/sqlalchemy/testing) directory?  For some reason I can't get those tests to run. The pytest command always reports `collected 0 items`.

**Reply by CaselIT** [reply]
regarding the test, you don't run them from `sqlalchemy.testing`.

have you read the dialect developers readme on how to run the suite tests on a dialect? https://github.com/sqlalchemy/sqlalchemy/blob/main/README.dialects.rst

**Comment by hab6** [comment]
Thanks for the help. I am fairly new to python and also had not seen the readme on developing new dialects. Will review and see if I can get the test suite running.

**Comment by hab6** [comment]
I am unable to get the dialect compliance suite to run and opened #11303 for help on that issue so as to not get too off track of the subject for this discussion (#11295).
